tools/src/wyvern/tools/parsing/coreparser/WyvernParser.jj

```diff
@@ -378,6 +380,29 @@ AST TypeDeclaration() :
   }
 }
 
+AST DatatypeDeclaration() :
+{ Token name; AST body = null; AST metadata = null; Type extendedType=null; List<Type> comprisesList=null; boolean isTagged = false;
+  Token selfName = null;
+  Object tagInfo = null; boolean isResource = false;
+}
+{
+  [<TAGGED> { isTagged = true; } ]
+  [<RESOURCE> { isResource = true; } ]
+  <DATATYPE> name=<IDENTIFIER>
+  [ <EXTENDS> extendedType=Type() ]
+  [ <COMPRISES> comprisesList=TypeList() ]
+  [ <LPAREN> selfName=<IDENTIFIER> <RPAREN> <ARROW> ]
```

Comment:
We probably don't need the ability to specify a selfName now either.  Though at some point, if we allow methods inside datatypes, then this would be useful.